File: src/dynamic-programming/minimum-falling-path-sum.ts
function minFallingPathSum(matrix: number[][]): number {
  const n = matrix.length
  const m = matrix.length
  //dp[i][j]代表从最顶层，下降到matrix[i][j]的最小值为dp[i][j]
  const dp: number[][] = []
  for (let i = 0; i < n; ++i) {
    dp[i] = []
  }
  for (let i = 0; i < m; ++i) {
    dp[0][i] = matrix[0][i]
  }

  for (let i = 1; i < n; ++i) {
    for (let j = 0; j < m; ++j) {
      dp[i][j] =
        Math.min(
          dp[i - 1][j - 1] ?? Infinity,
          dp[i - 1][j],
          dp[i - 1][j + 1] ?? Infinity
        ) + matrix[i][j]
    }
  }

  let ans = Infinity
  for (let i = 0; i < m; ++i) {
    ans = Math.min(ans, dp[n - 1][i])
  }

  return ans
}

File: src/dynamic-programming/minimum-falling-path-sum.cpp
class Solution {
 public:
  int minFallingPathSum(vector<vector<int>>& matrix) {
    int n = matrix.size();
    int m = matrix[0].size();

    for (int i = n - 2; i >= 0; --i) {
      for (int j = 0; j < m; ++j) {
        int min_num = matrix[i + 1][j];
        min_num = j - 1 >= 0 ? min(matrix[i + 1][j - 1], min_num) : min_num;

        min_num = j + 1 < m ? min(matrix[i + 1][j + 1], min_num) : min_num;

        matrix[i][j] += min_num;
      }
    }

    return *min_element(matrix[0].begin(), matrix[0].end());
  }
};

class Solution {
 public:
  int minFallingPathSum(vector<vector<int>>& matrix) {
    int n = matrix.size();
    vector<vector<int>> dp(n, vector<int>(n));
    copy(matrix[0].begin(), matrix[0].end(), dp[0].begin());
    for (int i = 1; i < n; i++) {
      for (int j = 0; j < n; j++) {
        int mn = dp[i - 1][j];
        if (j - 1 >= 0) {
          mn = min(mn, dp[i - 1][j - 1]);
        }
        if (j + 1 < n) {
          mn = min(mn, dp[i - 1][j + 1]);
        }
        dp[i][j] = mn + matrix[i][j];
      }
    }
    return *min_element(dp[n - 1].begin(), dp[n - 1].end());
  }
};